long-ride-alerts/src/main/java/org/apache/flink/training/exercises/longrides/LongRidesExercise.java

```diff
@@ -98,17 +102,62 @@ public static void main(String[] args) throws Exception {
     @VisibleForTesting
     public static class AlertFunction extends KeyedProcessFunction<Long, TaxiRide, Long> {
 
+        private static final long RIDE_DURATION_ALERT_TIME_IN_MS = 1000 * 60 * 60 * 2;
+        private ValueState<TaxiRide> startTaxiRide;
+        private ValueState<TaxiRide> endTaxiRide;
+
         @Override
-        public void open(Configuration config) throws Exception {
-            throw new MissingSolutionException();
+        public void open(Configuration config) {
+            startTaxiRide = getRuntimeContext().getState(new ValueStateDescriptor<>("startTaxiRide", TaxiRide.class));
+            endTaxiRide = getRuntimeContext().getState(new ValueStateDescriptor<>("endTaxiRide", TaxiRide.class));
         }
 
         @Override
         public void processElement(TaxiRide ride, Context context, Collector<Long> out)
-                throws Exception {}
+                throws Exception {
+            if (ride.isStart) {
+                TaxiRide taxiRide = endTaxiRide.value();
+                if (taxiRide != null) {
+                    handleRideDuration(ride, taxiRide, context, out);
+                    endTaxiRide.clear();
+                } else {
+                    startTaxiRide.update(ride);
+                    registerTimer(ride.getEventTimeMillis(), context);
+                }
+            } else {
+                TaxiRide taxiRide = startTaxiRide.value();
+                if (taxiRide != null) {
+                    handleRideDuration(taxiRide, ride, context, out);
+                    startTaxiRide.clear();
+                } else {
+                    endTaxiRide.update(ride);
+                }
+            }
+        }
+
+        private void handleRideDuration(TaxiRide startRide, TaxiRide endRide, Context context, Collector<Long> out) {
+            if (startRide != null && endRide != null) {
```

Comment:
I think you already checked both parameters not to be null before calling this method.

Reply:
done